### DISCUSSION TITLE
Support for `std::shared_ptr<const foo>`

### DISCUSSION THREAD

**Discussion by chrisrichardson** [wjakob/nanobind#235] [Q&A] [answered]
I want to wrap some C++ code that uses `std::shared_ptr<const foo>` but currently this is not working with nanobind (see below). I realise Python doesn't really have the concept of `const` but that makes it hard to wrap an existing C++ interface which does.
I thought maybe this could qualify as a "bug" but not sure if maybe more of a "feature request". The code below works OK without `const`.

```
#include <iostream>
#include <nanobind/nanobind.h>
#include <nanobind/stl/shared_ptr.h>

namespace nb = nanobind;

class MyObj {
public:
  MyObj() : value(1) {}

  int value;
};

void work(std::shared_ptr<const MyObj> w) { std::cout << w->value << "\n"; }

NB_MODULE(nanobind_example_ext, m) {

  nb::class_<MyObj>(m, "MyObj").def("__init__", [](MyObj *self) {
    new (self) MyObj();
  });
  m.def("work", &work);
}
```

**Comment by wjakob** [comment, marked as answer]
Supporting constant instances properly (Boost.Python-style) would be a massive change and is extremely unlikely to happen. Let me cite [the documentation](https://nanobind.readthedocs.io/en/latest/why.html): *The codebase has to adapt to the binding tool and not the other way around*

**Reply by chrisrichardson** [reply]
Thanks for the clear answer - I was couldn't find any specific information in the documentation, but I guess that general statement covers it.
